aiida/cmdline/utils/migration/v03_to_v04.py

```diff
@@ -0,0 +1,528 @@
+        unique_uuids = set(all_uuids)
+        if len(all_uuids) != len(unique_uuids):
+            echo.echo_critical("""{}s with exactly the same UUID found, cannot proceed further. Please contact AiiDA
+            developers: http://www.aiida.net/mailing-list/ to help you resolve this issue.""".format(entry_type))
+
+
+def migration_migrate_builtin_calculations(data):
+    """Apply migrations: 0019 - REV. 1.0.19
+    Remove 'simpleplugin' from ArithmeticAddCalculation and TemplatereplacerCalculation type
+
+    ATTENTION:
+
+    The 'process_type' column did not exist before migration 0010, consequently, it could not be present in any
+    export archive of the currently existing stable releases (0.12.*). Here, however, the migration acts
+    on the content of the 'process_type' column, which could only be introduced in alpha releases of AiiDA 1.0.
+    Assuming that 'add' and 'templateplacer' calculations are expected to have both 'type' and 'process_type' columns,
+    they will be added based solely on the 'type' column content (unlike the way it is done in the DB migration,
+    where the 'type_string' content was also checked).
+    """
+    for key, content in data['export_data']['Node'].items():
+        if 'type' in content:
+            if content['type'] == 'calculation.job.simpleplugins.arithmetic.add.ArithmeticAddCalculation.':
+                content['type'] = 'calculation.job.arithmetic.add.ArithmeticAddCalculation.'
+                content['process_type'] = 'aiida.calculations:arithmetic.add'
+            elif content['type'] == 'calculation.job.simpleplugins.templatereplacer.TemplatereplacerCalculation.':
+                content['type'] = 'calculation.job.templatereplacer.TemplatereplacerCalculation.'
+                content['process_type'] = 'aiida.calculations:templatereplacer'
+            elif content['type'] == 'data.code.Code.':
+                if data['node_attributes'][key]['input_plugin'] == 'simpleplugins.arithmetic.add':
+                    data['node_attributes'][key]['input_plugin'] = 'arithmetic.add'
+
+                elif data['node_attributes'][key]['input_plugin'] == 'simpleplugins.templatereplacer':
+                    data['node_attributes'][key]['input_plugin'] = 'templatereplacer'
+
+
+def migration_provenance_redesign(data):  # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    """Apply migration: 0020 - REV. 1.0.20
+    Provenance redesign
+    """
+    from aiida.manage.database.integrity.plugins import infer_calculation_entry_point
+    from aiida.manage.database.integrity import write_database_integrity_violation
+    from aiida.plugins.entry_point import ENTRY_POINT_STRING_SEPARATOR
+
+    fallback_cases = []
+    nodes_to_migrate = {}
+
+    for key, value in data['export_data']['Node'].items():
+        if 'type' in value and value['type'].startswith('calculation.job.'):
+            nodes_to_migrate[key] = value
+
+    if not nodes_to_migrate:
```

Comment:
This return here is probably wrong. There could be an archive without calculation jobs, but that still has other entries that would match all the conditions and migrations below. However, these would not be applied. Probably just remove this check/return and adapt the code below to be able to deal with an empty `nodes_to_migrate` (maybe even renaming to `calc_jobs_to_migrate` for clarity).

Reply:
Didn't all calculations start with the type `'calculation.job.'` or was it only `JobCalculation`?

Reply:
Never mind, I can see that further down we have `'calculation.inline'` as well.

Reply:
done